### PR TITLE
storage: Don't allow mounting over other mounted filesystems

### PR DESCRIPTION
**Pull request by mvollmer**
Demo: https://www.youtube.com/watch?v=tnf3Wl--Duw

### CODE REVIEW

File: pkg/storaged/fsys-tab.jsx [hunk 1]
```diff
@@ -95,18 +95,54 @@ export function find_blocks_for_mount_point(client, mount_point, self) {
...
+        const fs = client.blocks_fsys[p];
+
+        if (is_self(b))
+            continue;
+
+        if (fs) {
+            for (const mp of fs.MountPoints) {
+                const mpd = utils.decode_filename(mp);
+                if (mpd.length > mount_point.length && mpd.indexOf(mount_point) == 0 && mpd[mount_point.length] == "/")
+                    children[mpd] = b;
```
Comment (cockpituous) [inline pkg/storaged/fsys-tab.jsx:116]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230925-134144-be2f6638-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/fsys-tab.jsx [hunk 2]
```diff
@@ -95,18 +95,54 @@
...
                               other_blocks.map(nice_block_name).join(", "));
+
+    const children = find_children_for_mount_point(client, val, block);
+    if (Object.keys(children).length > 0)
+        return <>
+            {_("Filesystems are already mounted below this mountpoint.")}
+            <ul />
+            {Object.keys(children).map(m => <li key={m}>{nice_block_name(children[m])} on {m}</li>)}
+            <ul />
+            {_("Please unmount them first.")}
```
Comment (cockpituous) [inline pkg/storaged/fsys-tab.jsx:144]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230925-134144-be2f6638-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by mvollmer** [comment]
Oh, nice coverage report, very accurate!

File: pkg/storaged/fsys-tab.jsx [hunk 3]
```diff
@@ -99,14 +96,24 @@ function nice_block_name(block) {
...
         return cockpit.format(_("Mount point is already used for $0"),
                               other_blocks.map(nice_block_name).join(", "));
+
+    if (!ignore_overmounting) {
+        const children = utils.find_children_for_mount_point(client, val, block);
+        if (Object.keys(children).length > 0)
+            return <>
+                {_("Filesystems are already mounted below this mountpoint.")}
+                {Object.keys(children).map(m => <div key={m}>{cockpit.format("• $0 on $1", nice_block_name(children[m]), m)}</div>)}
+                {_("Please unmount them first.")}
```
Comment (cockpituous) [inline pkg/storaged/fsys-tab.jsx:114]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/nfs-details.jsx [hunk 4]
```diff
@@ -80,10 +80,10 @@ function get_exported_directories(server) {
 
 export function nfs_fstab_dialog(client, entry) {
     const mount_options = entry ? entry.fields[3] : "defaults";
-    const split_options = format.parse_options(mount_options == "defaults" ? "" : mount_options);
-    const opt_auto = !format.extract_option(split_options, "noauto");
-    const opt_ro = format.extract_option(split_options, "ro");
-    const extra_options = format.unparse_options(split_options);
+    const split_options = parse_options(mount_options == "defaults" ? "" : mount_options);
```
Comment (cockpituous) [inline pkg/storaged/nfs-details.jsx:83]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 5]
```diff
@@ -47,6 +47,88 @@ export function compare_versions(a, b) {
...
+    }
+}
+
+export function edit_crypto_config(block, modify) {
+    let old_config, new_config;
+
+    function commit() {
+        new_config[1]["track-parents"] = { t: 'b', v: true };
+        if (old_config)
+            return block.UpdateConfigurationItem(old_config, new_config, { });
```
Comment (cockpituous) [inline pkg/storaged/utils.js:79]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 6]
```diff
@@ -47,6 +47,88 @@
...
+        else
+            return block.AddConfigurationItem(new_config, { });
```
Comment (cockpituous) [inline pkg/storaged/utils.js:81]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 7]
```diff
@@ -47,6 +47,88 @@
...
+    }
+
+    return block.GetSecretConfiguration({}).then(
+        function (items) {
+            old_config = items.find(c => c[0] == "crypttab");
+            new_config = ["crypttab", old_config ? Object.assign({ }, old_config[1]) : { }];
```
Comment (cockpituous) [inline pkg/storaged/utils.js:87]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 8]
```diff
@@ -47,6 +47,88 @@
...
+            if (!new_config[1]['passphrase-contents'])
+                new_config[1]['passphrase-contents'] = { t: 'ay', v: encode_filename("") };
+
+            return modify(new_config[1], commit);
+        });
+}
+
+export function set_crypto_options(block, readonly, auto, nofail, netdev) {
+    return edit_crypto_config(block, (config, commit) => {
+        const opts = config.options ? parse_options(decode_filename(config.options.v)) : [];
```
Comment (cockpituous) [inline pkg/storaged/utils.js:102]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 9]
```diff
@@ -659,8 +741,34 @@
...
+        const fs = client.blocks_fsys[p];
+
+        if (is_self(b))
+            continue;
+
+        if (fs) {
+            for (const mp of fs.MountPoints) {
+                const mpd = decode_filename(mp);
+                if (mpd.length > mount_point.length && mpd.indexOf(mount_point) == 0 && mpd[mount_point.length] == "/")
+                    children[mpd] = b;
```
Comment (cockpituous) [inline pkg/storaged/utils.js:762]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 10]
```diff
@@ -682,16 +790,34 @@
             return actions;
         }
 
+        function enter_unmount(block, location, is_top) {
+            for (const u of usage) {
+                if (u.usage == 'mounted' && u.location == location) {
+                    if (is_top) {
+                        u.actions = get_actions(_("unmount"));
+                        u.set_noauto = false;
```
Comment (cockpituous) [inline pkg/storaged/utils.js:798]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 11]
```diff
@@ -682,16 +790,34 @@
...
+                    }
+                    return;
```
Comment (cockpituous) [inline pkg/storaged/utils.js:800]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 12]
```diff
@@ -682,16 +790,34 @@
...
+                }
+            }
+            usage.push({
+                level,
+                block,
+                usage: 'mounted',
+                location,
+                set_noauto: !is_top,
+                actions: is_top ? get_actions(_("unmount")) : [_("unmount")],
```
Comment (cockpituous) [inline pkg/storaged/utils.js:809]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 13]
```diff
@@ -682,16 +790,34 @@
...
-                    usage: 'mounted',
-                    block,
-                    location: decode_filename(mp),
-                    actions: get_actions(_("unmount")),
-                    blocking: false,
-                });
+                const mpd = decode_filename(mp);
+                const children = find_children_for_mount_point(client, mpd, null);
+                for (const c in children)
+                    enter_unmount(children[c], c, false);
```
Comment (cockpituous) [inline pkg/storaged/utils.js:819]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 14]
```diff
@@ -757,6 +884,37 @@
...
+    for (const conf of block.Configuration) {
+        if (conf[0] == "fstab" &&
+            decode_filename(conf[1].dir.v) == mount_point) {
+            const options = parse_options(get_block_mntopts(conf[1]));
+            if (options.indexOf("noauto") >= 0)
+                continue;
+            options.push("noauto");
+            const new_conf = [
+                "fstab",
+                Object.assign({ }, conf[1],
```
Comment (cockpituous) [inline pkg/storaged/utils.js:897]: These 11 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 15]
```diff
@@ -757,6 +884,37 @@
...
+                              {
+                                  opts: {
+                                      t: 'ay',
+                                      v: encode_filename(unparse_options(options))
```
Comment (cockpituous) [inline pkg/storaged/utils.js:901]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 16]
```diff
@@ -757,6 +884,37 @@
...
+                                  }
+                              })
+            ];
+            await block.UpdateConfigurationItem(conf, new_conf, { });
```
Comment (cockpituous) [inline pkg/storaged/utils.js:905]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 17]
```diff
@@ -757,6 +884,37 @@
...
+        }
+    }
+
+    const crypto_backing = client.blocks[block.CryptoBackingDevice];
+    if (crypto_backing) {
+        const crypto_backing_crypto = client.blocks_crypto[crypto_backing.path];
+        await set_crypto_auto_option(crypto_backing, false);
+        if (crypto_backing_crypto)
+            await crypto_backing_crypto.Lock({});
```
Comment (cockpituous) [inline pkg/storaged/utils.js:914]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 18]
```diff
@@ -767,10 +925,12 @@
...
 
-    function unmount(mounteds) {
-        return Promise.all(mounteds.map(m => {
-            return client.unmount_at(m.location, m.users);
-        }));
+    async function unmount(mounteds) {
+        for (const m of mounteds) {
+            await client.unmount_at(m.location, m.users);
+            if (m.set_noauto)
+                await set_fsys_noauto(client, m.block, m.location);
```
Comment (cockpituous) [inline pkg/storaged/utils.js:932]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230926-135906-360ec0bd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/fsys-tab.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/storaged/fsys-tab.jsx:114]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/nfs-details.jsx (hunk 4 above)
Comment (cockpituous) [inline pkg/storaged/nfs-details.jsx:83]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js (hunk 5 above)
Comment (cockpituous) [inline pkg/storaged/utils.js:79]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 19]
```diff
@@ -47,6 +47,88 @@ export function compare_versions(a, b) {
...
+
+export function edit_crypto_config(block, modify) {
+    let old_config, new_config;
+
+    function commit() {
+        new_config[1]["track-parents"] = { t: 'b', v: true };
+        if (old_config)
+            return block.UpdateConfigurationItem(old_config, new_config, { });
+        else
+            return block.AddConfigurationItem(new_config, { });
```
Comment (cockpituous) [inline pkg/storaged/utils.js:81]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 20]
```diff
@@ -47,6 +47,88 @@ export function compare_versions(a, b) {
...
+    }
+
+    return block.GetSecretConfiguration({}).then(
+        function (items) {
+            old_config = items.find(c => c[0] == "crypttab");
+            new_config = ["crypttab", old_config ? Object.assign({ }, old_config[1]) : { }];
```
Comment (cockpituous) [inline pkg/storaged/utils.js:87]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 21]
```diff
@@ -47,6 +47,88 @@ export function compare_versions(a, b) {
...
+            if (!new_config[1]['passphrase-contents'])
+                new_config[1]['passphrase-contents'] = { t: 'ay', v: encode_filename("") };
+
+            return modify(new_config[1], commit);
+        });
+}
+
+export function set_crypto_options(block, readonly, auto, nofail, netdev) {
+    return edit_crypto_config(block, (config, commit) => {
+        const opts = config.options ? parse_options(decode_filename(config.options.v)) : [];
```
Comment (cockpituous) [inline pkg/storaged/utils.js:102]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 22]
```diff
@@ -659,8 +741,34 @@ export function get_children(client, path) {
...
+        const fs = client.blocks_fsys[p];
+
+        if (is_self(b))
+            continue;
+
+        if (fs) {
+            for (const mp of fs.MountPoints) {
+                const mpd = decode_filename(mp);
+                if (mpd.length > mount_point.length && mpd.indexOf(mount_point) == 0 && mpd[mount_point.length] == "/")
+                    children[mpd] = b;
```
Comment (cockpituous) [inline pkg/storaged/utils.js:762]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 23]
```diff
@@ -682,16 +790,34 @@ export function get_active_usage(client, path, top_action, child_action) {
             return actions;
         }
 
+        function enter_unmount(block, location, is_top) {
+            for (const u of usage) {
+                if (u.usage == 'mounted' && u.location == location) {
+                    if (is_top) {
+                        u.actions = get_actions(_("unmount"));
+                        u.set_noauto = false;
```
Comment (cockpituous) [inline pkg/storaged/utils.js:798]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 24]
```diff
@@ -682,16 +790,34 @@ export function get_active_usage(client, path, top_action, child_action) {
...
+                    }
+                    return;
```
Comment (cockpituous) [inline pkg/storaged/utils.js:800]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 25]
```diff
@@ -682,16 +790,34 @@ export function get_active_usage(client, path, top_action, child_action) {
...
+                }
+            }
+            usage.push({
+                level,
+                block,
+                usage: 'mounted',
+                location,
+                set_noauto: !is_top,
+                actions: is_top ? get_actions(_("unmount")) : [_("unmount")],
```
Comment (cockpituous) [inline pkg/storaged/utils.js:809]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 26]
```diff
@@ -682,16 +790,34 @@ export function get_active_usage(client, path, top_action, child_action) {
...
-                    usage: 'mounted',
-                    block,
-                    location: decode_filename(mp),
-                    actions: get_actions(_("unmount")),
-                    blocking: false,
-                });
+                const mpd = decode_filename(mp);
+                const children = find_children_for_mount_point(client, mpd, null);
+                for (const c in children)
+                    enter_unmount(children[c], c, false);
```
Comment (cockpituous) [inline pkg/storaged/utils.js:819]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 27]
```diff
@@ -757,6 +884,37 @@ export function get_active_usage(client, path, top_action, child_action) {
...
+    for (const conf of block.Configuration) {
+        if (conf[0] == "fstab" &&
+            decode_filename(conf[1].dir.v) == mount_point) {
+            const options = parse_options(get_block_mntopts(conf[1]));
+            if (options.indexOf("noauto") >= 0)
+                continue;
+            options.push("noauto");
+            const new_conf = [
+                "fstab",
+                Object.assign({ }, conf[1],
```
Comment (cockpituous) [inline pkg/storaged/utils.js:897]: These 11 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 28]
```diff
@@ -757,6 +884,37 @@ export function get_active_usage(client, path, top_action, child_action) {
...
+                              {
+                                  opts: {
+                                      t: 'ay',
+                                      v: encode_filename(unparse_options(options))
```
Comment (cockpituous) [inline pkg/storaged/utils.js:901]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 29]
```diff
@@ -757,6 +884,37 @@ export function get_active_usage(client, path, top_action, child_action) {
...
+                                  }
+                              })
+            ];
+            await block.UpdateConfigurationItem(conf, new_conf, { });
```
Comment (cockpituous) [inline pkg/storaged/utils.js:905]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 30]
```diff
@@ -757,6 +884,37 @@ export function get_active_usage(client, path, top_action, child_action) {
...
+        }
+    }
+
+    const crypto_backing = client.blocks[block.CryptoBackingDevice];
+    if (crypto_backing) {
+        const crypto_backing_crypto = client.blocks_crypto[crypto_backing.path];
+        await set_crypto_auto_option(crypto_backing, false);
+        if (crypto_backing_crypto)
+            await crypto_backing_crypto.Lock({});
```
Comment (cockpituous) [inline pkg/storaged/utils.js:914]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 31]
```diff
@@ -767,10 +925,12 @@ export function teardown_active_usage(client, usage) {
...
 
-    function unmount(mounteds) {
-        return Promise.all(mounteds.map(m => {
-            return client.unmount_at(m.location, m.users);
-        }));
+    async function unmount(mounteds) {
+        for (const m of mounteds) {
+            await client.unmount_at(m.location, m.users);
+            if (m.set_noauto)
+                await set_fsys_noauto(client, m.block, m.location);
```
Comment (cockpituous) [inline pkg/storaged/utils.js:932]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-071616-0731117d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/nfs-details.jsx (hunk 4 above)
Comment (cockpituous) [inline pkg/storaged/nfs-details.jsx:83]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-091656-60b44b52-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js (hunk 5 above)
Comment (cockpituous) [inline pkg/storaged/utils.js:79]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-091656-60b44b52-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js (hunk 19 above)
Comment (cockpituous) [inline pkg/storaged/utils.js:81]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-091656-60b44b52-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js (hunk 20 above)
Comment (cockpituous) [inline pkg/storaged/utils.js:87]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-091656-60b44b52-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js (hunk 21 above)
Comment (cockpituous) [inline pkg/storaged/utils.js:102]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-091656-60b44b52-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js (hunk 23 above)
Comment (cockpituous) [inline pkg/storaged/utils.js:798]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-091656-60b44b52-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js (hunk 24 above)
Comment (cockpituous) [inline pkg/storaged/utils.js:800]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-091656-60b44b52-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 32]
```diff
@@ -764,16 +790,34 @@ export function get_active_usage(client, path, top_action, child_action) {
             return actions;
         }
 
+        function enter_unmount(block, location, is_top) {
+            for (const u of usage) {
+                if (u.usage == 'mounted' && u.location == location) {
+                    if (is_top) {
+                        u.actions = get_actions(_("unmount"));
+                        u.set_noauto = false;
```
Comment (cockpituous) [inline pkg/storaged/utils.js:798]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-101935-49dc2a8a-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 33]
```diff
@@ -764,16 +790,34 @@
...
+                    }
+                    return;
```
Comment (cockpituous) [inline pkg/storaged/utils.js:800]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230928-101935-49dc2a8a-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 34]
```diff
@@ -1178,7 +1186,24 @@ export const TeardownMessage = (usage) => {
...
-export function init_active_usage_processes(client, usage) {
+export function teardown_danger_message(usage, expect_single_unmount) {
+    if (is_expected_unmount(usage, expect_single_unmount))
+        return stop_processes_danger_message(usage[0].users);
+
+    const usage_with_users = usage.filter(u => u.users);
+    const n_processes = usage_with_users.reduce((sum, u) => sum + u.users.filter(u => u.pid).length, 0);
+    const n_services = usage_with_users.reduce((sum, u) => sum + u.users.filter(u => u.unit).length, 0);
+    if (n_processes > 0 && n_services > 0)
+        return _("Related processes and services will be forcefully stopped.");
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1197]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230929-123934-64a9fe0d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 35]
```diff
@@ -1178,7 +1186,24 @@
...
+    else if (n_processes > 0)
+        return _("Related processes will be forcefully stopped.");
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1199]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230929-123934-64a9fe0d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 36]
```diff
@@ -1178,7 +1186,24 @@
...
+    else if (n_services > 0)
+        return _("Related services will be forcefully stopped.");
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1201]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20230929-123934-64a9fe0d-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx (hunk 34 above)
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1197]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20231002-065009-8fd9cae1-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx (hunk 35 above)
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1199]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20231002-065009-8fd9cae1-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx (hunk 36 above)
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1201]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20231002-065009-8fd9cae1-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/utils.js [hunk 37]
```diff
@@ -849,10 +925,12 @@ export function teardown_active_usage(client, usage) {
...
     // condition upfront by reshuffling the data structures.
 
-    function unmount(mounteds) {
-        return Promise.all(mounteds.map(m => {
-            return client.unmount_at(m.location, m.users);
-        }));
+    async function unmount(mounteds) {
+        for (const m of mounteds) {
+            await client.unmount_at(m.location, m.users);
+            if (m.set_noauto)
```
Comment (jelly) [inline pkg/storaged/utils.js:931]: What is the rationale behind this change? If a user umount's he doesn't want to have a device auto mount anymore?

Comment (mvollmer) [reply pkg/storaged/utils.js:931]: Yes, when you unmount something with Cockpit, this means you don't want this to be mounted anymore, and rebooting shouldn't change that.

When you delete something with a filesystem on it, this will also unmount it, but it will immediately remove the fstab entry as well. So there is no need to add "noauto" to fstab during deletion.

However, when a filesystem needs to be unmounted because its parent is being deleted, then we _should_ add noauto to its fstab entry. (Since, of course, that entry will not be deleted.)

File: pkg/storaged/dialog.jsx [hunk 38]
```diff
@@ -1178,7 +1186,25 @@ export const TeardownMessage = (usage) => {
...
+
+    const usage_with_users = usage.filter(u => u.users);
+    const n_processes = usage_with_users.reduce((sum, u) => sum + u.users.filter(u => u.pid).length, 0);
+    const n_services = usage_with_users.reduce((sum, u) => sum + u.users.filter(u => u.unit).length, 0);
+    if (n_processes > 0 && n_services > 0) {
+        return _("Related processes and services will be forcefully stopped.");
+    } else if (n_processes > 0) {
+        return _("Related processes will be forcefully stopped.");
+    } else if (n_services > 0) {
+        return _("Related services will be forcefully stopped.");
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1201]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20231002-082429-e3e3b909-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1201]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20231002-114425-f2b455b6-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by jelly — APPROVED** [review]

File: pkg/storaged/dialog.jsx (hunk 38 above)
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1201]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19388-20231004-134539-0407a6e2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by jelly — APPROVED** [review]
